stores/utils.py

```diff
@@ -0,0 +1,179 @@
+                if variable_array.nchunks_initialized != variable_array.nchunks:
+                    errors.append(
+                        f'{variable} has {variable_array.nchunks - variable_array.nchunks_initialized} uninitialized chunks'
+                    )
+
+    if errors:
+        if raise_on_error:
+            raise ValueError(f'Found {len(errors)} errors: {errors}')
+        return False
+    return True
+
+
+def _pyramid_postprocess(
+    dt: dt.DataTree, levels: int, chunks: dict, pixels_per_tile: int
+) -> dt.DataTree:
+    '''Postprocess data pyramid
+
+    Adds multiscales metadata and sets Zarr encoding
+
+    Based on https://github.com/carbonplan/cmip6-downscaling
+
+    Parameters
+    ----------
+    dt : dt.DataTree
+        Input data pyramid
+    levels : int
+        Number of levels in pyramid
+    chunks : dict
+        Chunks for dims
+
+    Returns
+    -------
+    dt.DataTree
+        Updated data pyramid with metadata / encoding set
+    '''
+
+    for level in range(levels):
+        slevel = str(level)
+        dt.ds.attrs['multiscales'][0]['datasets'][level]['pixels_per_tile'] = pixels_per_tile
+
+        # set dataset chunks
+        dt[slevel].ds = dt[slevel].ds.chunk(chunks)
+
+        # set dataset encoding
+        dt[slevel].ds = set_web_zarr_encoding(
+            dt[slevel].ds, codec_config={'id': 'zlib', 'level': 1}, float_dtype='float32'
+        )
+        for var in ['time', 'time_bnds']:
+            if var in dt[slevel].ds:
+                dt[slevel].ds[var].encoding['dtype'] = 'int32'
+
+    # set global metadata
+    dt.ds.attrs.update({'title': 'multiscale data pyramid'})
+    return dt
+
+
+def pyramid(
+    ds_path: str, target: str, levels: int = 2, pixels_per_tile: int = 128, target_mb: int = 5
+) -> str:
+    '''Create a data pyramid from an xarray Dataset
+
+    Based on https://github.com/carbonplan/cmip6-downscaling
+
+    Parameters
+    ----------
+    ds_path : UPath
+        Path to input dataset
+    target : str
+        Path to write output data pyamid to
+    levels : int, optional
+        Number of levels in pyramid, by default 2
+
+    Returns
+    -------
+    target : str
+    '''
+
+    ds = xr.open_zarr(ds_path)
+
+    chunks = calc_chunk_dict(ds, target_mb=target_mb, pixels_per_tile=pixels_per_tile)
+
+    other_chunks = {'time': chunks['time']}
+
+    # create pyramid
+    dta = pyramid_regrid(
+        ds,
+        levels=levels,
+        pixels_per_tile=pixels_per_tile,
+        other_chunks=other_chunks,
+        regridder_kws={'ignore_degenerate': True},
+    )
+
+    dta = _pyramid_postprocess(dta, levels=levels, chunks=chunks, pixels_per_tile=pixels_per_tile)
+
+    # write to target
+    for child in dta.children.values():
+        for variable in child.ds.data_vars:
+            child[variable].encoding['write_empty_chunks'] = True
+
+    dta.to_zarr(target, mode='w')
+    validate_zarr_store(target)
+    return target
```

Comment:
As far as I know, most of this feature already exists in ndpyramid available at https://github.com/carbonplan/ndpyramid/blob/main/ndpyramid/utils.py.

Reply:
Thanks for pointing this out! Fixed!